ecs-agent/csiclient/csi_client.go

```diff
@@ -44,31 +57,85 @@ type csiClient struct {
 }
 
 // NewCSIClient creates a CSI client for the communication with CSI driver daemon.
-func NewCSIClient(socketIn string) CSIClient {
-	return &csiClient{csiSocket: socketIn}
+func NewCSIClient(socketIn string) csiClient {
+	return csiClient{csiSocket: socketIn}
 }
 
-// GetVolumeMetrics returns volume usage.
-func (cc *csiClient) GetVolumeMetrics(volumeId string, hostMountPath string) (*Metrics, error) {
-	// Set up a connection to the server
-	dialer := func(addr string, t time.Duration) (net.Conn, error) {
-		return net.Dial(PROTOCOL, addr)
+func (cc *csiClient) NodeStageVolume(ctx context.Context,
+	volID string,
+	publishContext map[string]string,
+	stagingTargetPath string,
+	fsType string,
+	accessMode v1.PersistentVolumeAccessMode,
+	secrets map[string]string,
+	volumeContext map[string]string,
+	mountOptions []string,
+	fsGroup *int64,
+) error {
+	conn, err := cc.grpcDialConnect()
+	if err != nil {
+		logger.Error("NodeStage: CSI Connection Error", logger.Fields{
+			field.Error: err,
+		})
+		return err
 	}
-	conn, err := grpc.Dial(
-		cc.csiSocket,
-		grpc.WithTransportCredentials(insecure.NewCredentials()),
-		grpc.WithDialer(dialer),
-	)
+	defer conn.Close()
+
+	client := csi.NewNodeClient(conn)
+
+	defaultVolumeCapability := &csi.VolumeCapability{
+		AccessType: &csi.VolumeCapability_Mount{
+			Mount: &csi.VolumeCapability_MountVolume{},
+		},
+		AccessMode: &csi.VolumeCapability_AccessMode{
+			Mode: csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER,
+		},
+	}
+	req := csi.NodeStageVolumeRequest{
+		VolumeId:          volID,
+		PublishContext:    publishContext,
+		StagingTargetPath: stagingTargetPath,
+		VolumeCapability:  defaultVolumeCapability,
+		Secrets:           secrets,
+		VolumeContext:     volumeContext,
+	}
+
+	if fsType == fsTypeBlockName {
+		req.VolumeCapability.AccessType = &csi.VolumeCapability_Block{
+			Block: &csi.VolumeCapability_BlockVolume{},
+		}
+	} else {
+		mountVolume := &csi.VolumeCapability_MountVolume{
+			FsType:     fsType,
+			MountFlags: mountOptions,
+		}
+		req.VolumeCapability.AccessType = &csi.VolumeCapability_Mount{
+			Mount: mountVolume,
+		}
+	}
+
+	_, err = client.NodeStageVolume(ctx, &req)
+
 	if err != nil {
-		logger.Error("Error building a connection to CSI driver", logger.Fields{
+		logger.Error("Error staging volume via CSI driver", logger.Fields{
 			field.Error: err,
-			"Socket":    cc.csiSocket,
 		})
+		return err
+	}
+	return nil
+}
+
+// GetVolumeMetrics returns volume usage.
+func (cc *csiClient) GetVolumeMetrics(volumeId string, hostMountPath string) (*Metrics, error) {
+	conn, err := cc.grpcDialConnect()
+	if err != nil {
+		logger.Error("GetVolumeMetrics: CSI Connection Error")
```

Comment:
nit(non-blocking): perhaps we should also include the error in the log statement?

Reply:
done